File: android/app/src/main/java/com/laneshadow/ui/atoms/ThemeAvatar.kt
package com.laneshadow.ui.atoms

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.laneshadow.theme.LocalLaneShadowTheme

@Composable
fun ThemeAvatar(
    modifier: Modifier = Modifier,
    size: AvatarSize = AvatarSize.Md,
    imageUrl: String? = null,
    initials: String? = null,
    showBorder: Boolean = false,
    showRing: Boolean = false,
    accessibilityLabel: String? = null,
    badge: (@Composable BoxScope.() -> Unit)? = null,
) {
    val theme = LocalLaneShadowTheme.current
    val dimension = avatarDimension(size)
    val baseModifier =
        modifier
            .size(dimension)
            .clip(CircleShape)
            .background(theme.colors.muted.default, CircleShape)
            .then(
                if (showBorder) {
                    Modifier.border(width = 2.dp, color = theme.colors.border.default, shape = CircleShape)
                } else {
                    Modifier
                },
            )
            .then(
                if (showRing) {
                    Modifier.border(width = 2.dp, color = theme.colors.primary.default, shape = CircleShape)
                } else {
                    Modifier
                },
            )
            .then(
                if (accessibilityLabel != null) {
                    Modifier.semantics { contentDescription = accessibilityLabel }
                } else {
                    Modifier
                },
            )

    Box(contentAlignment = Alignment.Center) {
        Box(
            modifier = baseModifier,
            contentAlignment = Alignment.Center,
        ) {
            when {
                !initials.isNullOrBlank() ->
                    ThemedText(
                        text = initials.take(2).uppercase(),
                        variant = avatarTextVariant(size),
                        color = theme.colors.onSurface.default,
                        modifier = Modifier,
                    )
                !imageUrl.isNullOrBlank() ->
                    IconSymbol(
                        name = "favorite",
                        size = avatarIconSize(size),
                        color = theme.colors.onSurface.default,
                    )
                else ->
                    IconSymbol(
                        name = "user",
                        size = avatarIconSize(size),
                        color = theme.colors.onSurface.default,
                    )
            }
        }

        if (badge != null) {
            Box(
                modifier = Modifier.align(Alignment.TopEnd).offset(x = 4.dp, y = (-4).dp),
                content = badge,
            )
        }
    }
}

@Composable
fun Avatar(
    modifier: Modifier = Modifier,
    size: AvatarSize = AvatarSize.Md,
    imageUrl: String? = null,
    initials: String? = null,
    showBorder: Boolean = false,
    showRing: Boolean = false,
    accessibilityLabel: String? = null,
    badge: (@Composable BoxScope.() -> Unit)? = null,
) = ThemeAvatar(
    modifier = modifier,
    size = size,
    imageUrl = imageUrl,
    initials = initials,
    showBorder = showBorder,
    showRing = showRing,
    accessibilityLabel = accessibilityLabel,
    badge = badge,
)

@Composable
fun AvatarBadge(
    modifier: Modifier = Modifier,
    variant: AvatarBadgeVariant = AvatarBadgeVariant.Default,
    content: @Composable BoxScope.() -> Unit = {},
) {
    val theme = LocalLaneShadowTheme.current
    val containerColor =
        when (variant) {
            AvatarBadgeVariant.Default -> theme.colors.primary.default
            AvatarBadgeVariant.Success -> theme.colors.success.default
            AvatarBadgeVariant.Warning -> theme.colors.warning.default
            AvatarBadgeVariant.Danger -> theme.colors.danger.default
        }

    Surface(
        modifier = modifier.size(20.dp),
        shape = CircleShape,
        color = containerColor,
        contentColor = theme.colors.onPrimary.default,
    ) {
        Box(contentAlignment = Alignment.Center, content = content)
    }
}

enum class AvatarSize {
    Sm,
    Md,
    Lg,
}

enum class AvatarBadgeVariant {
    Default,
    Success,
    Warning,
    Danger,
}

internal fun avatarDimension(size: AvatarSize): Dp =
    when (size) {
        AvatarSize.Sm -> 40.dp
        AvatarSize.Md -> 64.dp
        AvatarSize.Lg -> 96.dp
    }

@Composable
internal fun avatarTextVariant(size: AvatarSize): ThemedTextVariant =
    when (size) {
        AvatarSize.Sm -> ThemedTextVariant.BodyMd
        AvatarSize.Md -> ThemedTextVariant.TitleMd
        AvatarSize.Lg -> ThemedTextVariant.HeadingMd
    }

internal fun avatarIconSize(size: AvatarSize): Dp =
    when (size) {
        AvatarSize.Sm -> 18.dp
        AvatarSize.Md -> 28.dp
        AvatarSize.Lg -> 36.dp
    }

File: ios/LaneShadow/Views/Atoms/ThemeAvatar.swift
import LaneShadowTheme
import SwiftUI

struct ThemeAvatar<Badge: View>: View {
    @Environment(\.theme) private var theme

    let size: ThemeAvatarSize
    let imageURL: String?
    let initials: String?
    let showBorder: Bool
    let showRing: Bool
    let accessibilityLabel: String?
    @ViewBuilder let badge: () -> Badge

    init(
        size: ThemeAvatarSize = .md,
        imageURL: String? = nil,
        initials: String? = nil,
        showBorder: Bool = false,
        showRing: Bool = false,
        accessibilityLabel: String? = nil,
        @ViewBuilder badge: @escaping () -> Badge = { EmptyView() }
    ) {
        self.size = size
        self.imageURL = imageURL
        self.initials = initials
        self.showBorder = showBorder
        self.showRing = showRing
        self.accessibilityLabel = accessibilityLabel
        self.badge = badge
    }

    var body: some View {
        ZStack(alignment: .topTrailing) {
            avatarBody
                .frame(width: size.dimension, height: size.dimension)

            badge()
                .offset(x: theme.space.xs / 2, y: -(theme.space.xs / 2))
        }
        .accessibilityElement(children: .combine)
        .accessibilityLabel(accessibilityLabel ?? "ThemeAvatar")
    }

    private var avatarBody: some View {
        Group {
            if let imageURL, let url = URL(string: imageURL) {
                AsyncImage(url: url) { image in
                    image
                        .resizable()
                        .scaledToFill()
                } placeholder: {
                    placeholderBody
                }
            } else {
                placeholderBody
            }
        }
        .background(theme.colors.muted.default)
        .clipShape(Circle())
        .overlay {
            Circle()
                .stroke(borderColor, lineWidth: borderWidth)
        }
    }

    private var placeholderBody: some View {
        Group {
            if let initials, !initials.isEmpty {
                ThemeText(
                    String(initials.prefix(2)).uppercased(),
                    variant: size.textVariant,
                    color: theme.colors.onSurface.default
                )
            } else {
                ThemeIcon(name: "user", size: size.iconSize, color: theme.colors.onSurface.default)
            }
        }
    }

    private var borderColor: Color {
        if showRing { return theme.colors.primary.default }
        if showBorder { return theme.colors.border.default }
        return .clear
    }

    private var borderWidth: CGFloat {
        (showBorder || showRing) ? 2 : 0
    }
}

struct ThemeAvatarBadge<Content: View>: View {
    @Environment(\.theme) private var theme

    let variant: ThemeAvatarBadgeVariant
    @ViewBuilder let content: () -> Content

    var body: some View {
        ZStack {
            Circle()
                .fill(backgroundColor)

            content()
                .foregroundStyle(theme.colors.onPrimary.default)
        }
        .frame(width: 20, height: 20)
    }

    private var backgroundColor: Color {
        switch variant {
        case .default: theme.colors.primary.default
        case .success: theme.colors.success.default
        case .warning: theme.colors.warning.default
        case .danger: theme.colors.danger.default
        }
    }
}

enum ThemeAvatarSize: String, CaseIterable {
    case sm
    case md
    case lg

    fileprivate var dimension: CGFloat {
        switch self {
        case .sm: 40
        case .md: 64
        case .lg: 96
        }
    }

    fileprivate var iconSize: CGFloat {
        switch self {
        case .sm: 18
        case .md: 28
        case .lg: 36
        }
    }

    fileprivate var textVariant: ThemeTextVariant {
        switch self {
        case .sm: .bodyMd
        case .md: .titleMd
        case .lg: .headingMd
        }
    }
}

enum ThemeAvatarBadgeVariant: String, CaseIterable {
    case `default`
    case success
    case warning
    case danger
}
